Description: make shortcuts work for platform menu bars
 When a platform menu bar is used, the QMenuBar is hidden, so shortcuts
 for QActions attached only to it do not work.
Origin: upstream, https://code.qt.io/cgit/qt/qtbase.git/commit/?id=287f548d4c7cc594
Bug: https://bugs.launchpad.net/bugs/1380702
Last-Update: 2017-01-25

--- a/src/widgets/kernel/qshortcut.cpp
+++ b/src/widgets/kernel/qshortcut.cpp
@@ -141,9 +141,11 @@
 static bool correctWidgetContext(Qt::ShortcutContext context, QWidget *w, QWidget *active_window)
 {
     bool visible = w->isVisible();
-#ifdef Q_OS_MAC
-    if (!qApp->testAttribute(Qt::AA_DontUseNativeMenuBar) && qobject_cast<QMenuBar *>(w))
-        visible = true;
+#ifndef QT_NO_MENUBAR
+    if (QMenuBar *menuBar = qobject_cast<QMenuBar *>(w)) {
+        if (menuBar->isNativeMenuBar())
+            visible = true;
+    }
 #endif
 
     if (!visible || !w->isEnabled())
--- a/src/widgets/widgets/qmenubar.cpp
+++ b/src/widgets/widgets/qmenubar.cpp
@@ -1288,10 +1288,12 @@
     } else if(e->type() == QEvent::ActionRemoved) {
         e->action()->disconnect(this);
     }
-    if (isVisible()) {
+    // updateGeometries() is also needed for native menu bars because
+    // it updates shortcutIndexMap
+    if (isVisible() || isNativeMenuBar())
         d->updateGeometries();
+    if (isVisible())
         update();
-    }
 }
 
 /*!
@@ -1708,6 +1710,13 @@
 {
     Q_Q(QMenuBar);
     QAction *act = actions.at(id);
+    if (act && act->menu()) {
+        if (QPlatformMenu *platformMenu = act->menu()->platformMenu()) {
+            platformMenu->showPopup(q->windowHandle(), actionRects.at(id), Q_NULLPTR);
+            return;
+        }
+    }
+
     setCurrentAction(act, true, true);
     if (act && !act->menu()) {
         activateAction(act, QAction::Trigger);
